http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/NettyRequestExecutor.java

```diff
@@ -35,91 +40,127 @@
 import software.amazon.awssdk.http.Protocol;
-import software.amazon.awssdk.http.async.AbortableRunnable;
 import software.amazon.awssdk.http.nio.netty.internal.http2.Http2ToHttpInboundAdapter;
 import software.amazon.awssdk.http.nio.netty.internal.http2.HttpToHttp2OutboundAdapter;
 import software.amazon.awssdk.http.nio.netty.internal.utils.ChannelUtils;
-import software.amazon.awssdk.utils.FunctionalUtils.UnsafeRunnable;
 
 @SdkInternalApi
-public final class RunnableRequest implements AbortableRunnable {
-
-    private static final Logger log = LoggerFactory.getLogger(RunnableRequest.class);
+public final class NettyRequestExecutor {
+    private static final Logger log = LoggerFactory.getLogger(NettyRequestExecutor.class);
+    private static final RequestAdapter REQUEST_ADAPTER = new RequestAdapter();
+    private static final AtomicLong EXECUTION_COUNTER = new AtomicLong(0L);
+    private final long executionId = EXECUTION_COUNTER.incrementAndGet();
     private final RequestContext context;
-    private volatile Channel channel;
+    private CompletableFuture<Void> executeFuture;
+    private Channel channel;
 
-    public RunnableRequest(RequestContext context) {
+    public NettyRequestExecutor(RequestContext context) {
         this.context = context;
     }
 
-    @Override
-    public void run() {
-        context.channelPool().acquire().addListener((Future<Channel> channelFuture) -> {
-            if (channelFuture.isSuccess()) {
-                try {
-                    channel = channelFuture.getNow();
-                    channel.attr(REQUEST_CONTEXT_KEY).set(context);
-                    channel.attr(RESPONSE_COMPLETE_KEY).set(false);
-                    makeRequest(context.nettyRequest());
-                } catch (Exception e) {
-                    handleFailure(() -> "Failed to make request to " + endpoint(), e);
-                }
-            } else {
-                handleFailure(() -> "Failed to create connection to " + endpoint(), channelFuture.cause());
+    @SuppressWarnings("unchecked")
+    public CompletableFuture<Void> execute() {
+        Future<Channel> channelFuture = context.channelPool().acquire();
+        executeFuture = createExecuteFuture(channelFuture);
+        channelFuture.addListener((GenericFutureListener) this::makeRequestListener);
+        return executeFuture;
+    }
+
+    /**
+     * Convenience method to create the execution future and set up the cancellation logic.
+     *
+     * @param channelFuture The Netty future holding the channel.
+     *
+     * @return The created execution future.
+     */
+    private CompletableFuture<Void> createExecuteFuture(Future<Channel> channelFuture) {
+        final CompletableFuture<Void> future = new CompletableFuture<>();
+
+        future.whenComplete((r, t) -> {
+            if (t == null) {
+                return;
             }
+
+            channelFuture.addListener((Future<Channel> f) -> {
```

Comment:
Not necessarily but it's true the other way around: if the future is cancelled, then it's completed exceptionally. There are other cases (like in `handleFailure`) where we complete the future exceptionally and want to close the connection so that handles that here as well.